src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+ * Increments the progress of the objective by the specified amount, towards the goal.
+ * @param sd      as a pointer to map_session_data
+ * @param ad      as a pointer to the achievement_data
+ * @param obj_idx as the index of the objective.
+ * @param progress   as the progress of the objective to be added.
+ */
+void achievement_progress_add(struct map_session_data *sd, struct achievement_data *ad, const unsigned int obj_idx, const int progress)
+{
+	struct achievement *ach = NULL;
+
+	nullpo_retv(sd);
+	nullpo_retv(ad);
+
+	Assert_retv(progress > 0);
+	Assert_retv(obj_idx < VECTOR_LENGTH(ad->objective));
+
+	// find existing or create a new entry.
+	ach = achievement->find_or_create(sd, ad->id, true);
+
+	if (ach->completed_at)
+		return;
+	
+	// Check and increment the objective count.
+	if (!ach->objective[obj_idx] || ach->objective[obj_idx] < VECTOR_INDEX(ad->objective, obj_idx).goal) {
+		ach->objective[obj_idx] = min(progress + ach->objective[obj_idx], VECTOR_INDEX(ad->objective, obj_idx).goal);
+
+		// Check if the Achievement is complete.
+		if (achievement->check_complete(sd, ad)) {
+			achievement->validate_achieve(sd, ad->id);
+			ach->completed_at = time(NULL);
+		}
+
+		// update client.
+		clif->achievement_send_update(sd->fd, sd, ad);
+	}
+}
+
+/**
+ * Compare an absolute progress value against the goal of an objective.
+ * Does not add/increase progress.
+ * @param sd        pointer to map-server session data.
+ * @param ad        pointer to achievement data.
+ * @param obj_idx   index of the objective in question.
+ * @param progress  progress of the objective in question.
+ */
+void achievement_progress_set(struct map_session_data *sd, struct achievement_data *ad, const unsigned int obj_idx, const int progress)
+{
+	struct achievement *ach = NULL;
+
+	nullpo_retv(sd);
+	nullpo_retv(ad);
+
+	Assert_retv(progress > 0);
+	Assert_retv(obj_idx < VECTOR_LENGTH(ad->objective));
+
+	if (progress >= VECTOR_INDEX(ad->objective, obj_idx).goal) {
+		if ((ach = achievement->find_or_create(sd, ad->id, true)) && ach->completed_at)
+			return;
+
+		ach->objective[obj_idx] = VECTOR_INDEX(ad->objective, obj_idx).goal;
+
+		if (achievement->check_complete(sd, ad)) {
+			achievement->validate_achieve(sd, ad->id);
+			ach->completed_at = time(NULL);
+		}
+
+		clif->achievement_send_update(sd->fd, sd, ad);
+	}
+}
+
+/**
+* Checks if the given criteria satisfies the achievement's objective.
+* @param objective pointer to the achievement's objectives data.
+* @param criteria  pointer to the current session's criteria as a comparand.
+* @return true if all criteria are satisfied, else false.
+*/
+bool achievement_check_criteria(const struct achievement_objective *objective, const struct achievement_objective *criteria)
+{
+	bool equal = true;
+	int i = 0, j = 0;
+
+	nullpo_retr(false, objective);
+	nullpo_retr(false, criteria);
+
+	/* Monster Id */
+	equal = (objective->mobid && objective->mobid == criteria->mobid);
```

Comment:
In the case where one objective has more than one criteria for example - mobid and itemid, your suggestion would return false every time without checking the second criteria.